src/librustc_errors/annotate_rs_emitter.rs

```diff
@@ -0,0 +1,213 @@
+            let primary_lo = if let Some(ref primary_span) =
+                self.msp.primary_span().as_ref() {
+                source_map.lookup_char_pos(primary_span.lo())
+            } else {
+                // FIXME(#59346): Not sure when this is the case and what
+                // should be done if it happens
+                return None
+            };
+            let annotated_files = FileWithAnnotatedLines::collect_annotations(
+                &self.msp,
+                &self.source_map
+            );
+            let slices = self.slices_for_files(annotated_files, primary_lo);
+
+            Some(Snippet {
+                title: Some(Annotation {
+                    label: Some(self.message.to_string()),
+                    id: self.code.clone().map(|c| {
+                        match c {
+                            DiagnosticId::Error(val) | DiagnosticId::Lint(val) => val
+                        }
+                    }),
+                    annotation_type: Self::annotation_type_for_level(self.level),
+                }),
+                footer: vec![],
+                slices: slices,
+            })
+        } else {
+            // FIXME(#59346): Is it ok to return None if there's no source_map?
+            None
+        }
+    }
+
+    fn slices_for_files(
+        &self,
+        annotated_files: Vec<FileWithAnnotatedLines>,
+        primary_lo: Loc
+    ) -> Vec<Slice> {
+        // FIXME(#59346): Provide a test case where `annotated_files` is > 1
+        annotated_files.iter().flat_map(|annotated_file| {
+            annotated_file.lines.iter().map(|line| {
+                let line_source = Self::source_string(annotated_file.file.clone(), &line);
+                Slice {
+                    source: line_source,
+                    line_start: line.line_index,
+                    origin: Some(primary_lo.file.name.to_string()),
+                    // FIXME(#59346): Not really sure when `fold` should be true or false
+                    fold: false,
+                    annotations: line.annotations.iter().map(|a| {
+                        self.annotation_to_source_annotation(a.clone())
+                    }).collect(),
+                }
+            }).collect::<Vec<Slice>>()
+        }).collect::<Vec<Slice>>()
+    }
+
+    /// Turns a `crate::snippet::Annotation` into a `SourceAnnotation`
+    fn annotation_to_source_annotation(
+        &self,
+        annotation: crate::snippet::Annotation
+    ) -> SourceAnnotation {
+        SourceAnnotation {
+            range: (annotation.start_col, annotation.end_col),
+            label: annotation.label.unwrap_or("".to_string()),
+            annotation_type: Self::annotation_type_for_level(self.level)
+        }
+    }
+
+    /// Provides the source string for the given `line` of `file`
+    fn source_string(file: Lrc<SourceFile>,
+                     line: &Line) -> String {
+        let source_string = match file.get_line(line.line_index - 1) {
```

Comment:
I think this function's entire body can be `file.get_line(line.line_index - 1).map(|s| s.to_string()).unwrap_or_default()`, but I'm not sure what `get_line` returns.